File: app/src/main/java/com/ternaryop/utils/drawer/adapter/DrawerAdapter.java
package com.ternaryop.utils.drawer.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.ternaryop.utils.R;

public class DrawerAdapter extends ArrayAdapter<DrawerItem> {
    private static LayoutInflater inflater = null;
    private boolean isSelectionEnabled;

    public DrawerAdapter(Context context) {
        super(context, 0);
        inflater = LayoutInflater.from(context);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        DrawerItem item = getItem(position);
        ViewHolder holder;

        if (convertView == null) {
            if (item.isHeader()) {
                convertView = inflater.inflate(R.layout.drawer_header_list_item, parent, false);
            } else if (item.isDivider()) {
                convertView = inflater.inflate(R.layout.drawer_divider_list_item, parent, false);
            } else {
                convertView = inflater.inflate(R.layout.drawer_list_item, parent, false);
            }
            holder = new ViewHolder(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }
        // the header hasn't title
        if (holder.title != null) {
            holder.title.setText(item.getTitle());
        }
        // the header hasn't counter
        if (holder.counter != null) {
            holder.counter.setVisibility(View.GONE);
            if (item.isCounterVisible() && item.getCountRetriever() != null) {
                item.getCountRetriever().updateCount(holder.counter);
            }
        }
        return convertView;
    }

    @Override
    public boolean isEnabled(int position) {
        return isSelectionEnabled() && !getItem(position).isHeader();
    }

    @Override
    public int getItemViewType(int position) {
        DrawerItem drawerItem = super.getItem(position);
        if (drawerItem.isHeader()) {
            return 0;
        } else if (drawerItem.isDivider()) {
            return 1;
        }
        return 2;
    }

    @Override
    public int getViewTypeCount() {
        // header, divider and normal row
        return 3;
    }

    public boolean isSelectionEnabled() {
        return isSelectionEnabled;
    }

    public void setSelectionEnabled(boolean isSelectionEnabled) {
        this.isSelectionEnabled = isSelectionEnabled;
    }

    private class ViewHolder {
        final TextView title;
        final TextView counter;

        ViewHolder(View view) {
            title = (TextView) view.findViewById(android.R.id.text1);
            counter = (TextView) view.findViewById(R.id.counter);
        }
    }
}

File: app/src/main/java/com/ternaryop/utils/drawer/adapter/DrawerAdapter.kt
package com.ternaryop.utils.drawer.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import com.ternaryop.utils.R

class DrawerAdapter(context: Context) : ArrayAdapter<DrawerItem>(context, 0) {
    var isSelectionEnabled: Boolean = false
    private var inflater = LayoutInflater.from(context)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val inflatedView: View
        val item = getItem(position)!!
        val holder: ViewHolder

        if (convertView == null) {
            inflatedView = when {
                item.isHeader -> inflater!!.inflate(R.layout.drawer_header_list_item, parent, false)
                item.isDivider -> inflater!!.inflate(R.layout.drawer_divider_list_item, parent, false)
                else -> inflater!!.inflate(R.layout.drawer_list_item, parent, false)
            }
            holder = ViewHolder(inflatedView)
            inflatedView.tag = holder
        } else {
            inflatedView = convertView
            holder = inflatedView.tag as ViewHolder
        }
        // the header hasn't title
        holder.title?.text = item.title

        // the header hasn't counter
        holder.counter?.let { counter ->
            counter.visibility = View.GONE
            if (item.isCounterVisible) {
                if (item.badge != null && item.badge!!.isNotBlank()) {
                    counter.text = item.badge.toString()
                    counter.visibility = View.VISIBLE
                }
            }
        }
        return inflatedView
    }

    override fun isEnabled(position: Int): Boolean {
        return isSelectionEnabled && !getItem(position)!!.isHeader
    }

    override fun getItemViewType(position: Int): Int {
        val drawerItem = super.getItem(position)
        if (drawerItem!!.isHeader) {
            return 0
        } else if (drawerItem.isDivider) {
            return 1
        }
        return 2
    }

    override fun getViewTypeCount(): Int {
        // header, divider and normal row
        return 3
    }

    private inner class ViewHolder internal constructor(view: View) {
        val title = view.findViewById<View>(android.R.id.text1) as? TextView
        val counter = view.findViewById<View>(R.id.counter) as? TextView
    }

    fun getItemById(id: Int): DrawerItem? {
        for (i in 0 until count) {
            val item = getItem(i)!!

            if (item.itemId == id) {
                return item
            }
        }
        return null
    }
}
